### DISCUSSION TITLE
Next.js 13: How can I access the params for GET in route.ts?

### DISCUSSION THREAD

**Discussion by dapjeongneo** [vercel/next.js#47072] [Help] [answered]
### Summary

I have a function in my client to call an API in route.ts.

How can I access the params `filename` in the `GET` function? Apparently the `filename` shows up in the debugger under the `request.url` instead of `params`. `params` is undefined. Can someone please help me out? Thank you in advance.

Extra credit: Is it possible to pass an image `HTMLInputElement.files` from the `button.tsx` to the `route.ts` so that I can upload an image directly to S3 in `route.ts`? My current approach is to get an S3 URL back from the `GET` and upload in the `button.tsx`.

### Additional information

```js
// button.tsx
const res = await axios.get(`/path/to/route`, {
  params: {
    filename: file.name,
  },
})

// route.ts
export async function GET(
  request: NextRequest,
  { params }: { params: { filename: string } }
) {
```

**Comment by icyJoseph** [comment, marked as answer]
Have you tried,

```js
request.nextUrl.searchParams
```

- https://beta.nextjs.org/docs/api-reference/request#nexturl

To be more precise:

```ts
import { NextRequest } from "next/server";

export async function GET(request: NextRequest) {
  console.log(request.nextUrl.searchParams.get("foo"));
  return new Response("Hello, Next.js!");
}
```

Which would log `bar` for this request:

```http
GET http://localhost:3000/api/hello?foo=bar
```

**Reply by dapjeongneo** [reply]
yes, it works! thanks, icyJoseph! do you know how to make the `context` work [here](https://beta.nextjs.org/docs/api-reference/file-conventions/route#context-optional)?

**Reply by icyJoseph** [reply]
What do you want out of the context?

**Reply by dapjeongneo** [reply]
im not looking for anything in particular, but i wonder how can i make use of it. do you know?

**Reply by icyJoseph** [reply]
Aha, those are for dynamic paths, like `/users/[id]/user` then that context object contains the value of `id`

**Reply by dapjeongneo** [reply]
gotcha, thanks! :)

**Reply by imheldermartins** [reply]
Thanks dude! It's working for me yet. 
Next version: 13.4.19

**Reply by bsilva0x87** [reply]
> 
Same here, @heldermartins4! I actually found a solution in the upcoming release, version 13.4.19.

```JS
export const GET = async (req, { params }) {
    // params object contains the request id param { id: 'the-request-id-param-here'}...
    const {id = "" } = params; 
    ...
}

**Reply by MoonSangJin** [reply]
thank you so much

**Reply by ghost** [reply]
Thanks bsilva I have been banging my head all day on this. Can you post the URL to the document you read?

**Reply by landsman** [reply]
I had to search this on GitHub, it is not easy to find in web documentation

**Reply by sajid-munawar** [reply]
https://nextjs.org/docs/app/api-reference/functions/next-request#nexturl

**Reply by Muneeb-Mohd** [reply]
Worked like a charm, would like to know what params logs out though, seriously curious

Never mind, it basically logs whatever is used as dynamic value.

**Reply by jericrealubit** [reply]
very cool! Thank you.

**Reply by nicope** [reply]
you're now my new god

**Reply by alsgy2001** [reply]
> Have you tried,
> 
> ```js
> request.nextUrl.searchParams
> ```
> 
> * https://beta.nextjs.org/docs/api-reference/request#nexturl
> 
> To be more precise:
> 
> ```ts
> import { NextRequest } from "next/server";
> 
> export async function GET(request: NextRequest) {
>   console.log(request.nextUrl.searchParams.get("foo"));
>   return new Response("Hello, Next.js!");
> }
> ```
> 
> Which would log `bar` for this request:
> 
> ```httpspec
> GET http://localhost:3000/api/hello?foo=bar
> ```

It worked for me.

**Comment by AnsonZnl** [comment]
very thanks

**Comment by sajid-munawar** [comment]
Thank you so much

**Comment by tjcchen** [comment]
works like a charm.

**Comment by laubonghaudoi** [comment]
The [official doc](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#convention) isn't very helpful, I managed to get the params with this

```ts
export async function GET(request: Request) {
  const { searchParams } = new URL(request.url);
  const my_param = searchParams.get("my_param");
)
```
The top answer uses `import { NextRequest } from "next/server";` which is Next's extension. If you are opting out of caching like me using `req: Request`, your `request` won't have `nextUrl`. So you have to `new URL(request.url)` to get the params.

**Reply by emporteme** [reply]
Nice

**Comment by tuncatunc** [comment]
Nextjs 13, to get the dynamic route parameters from a an API route`/api/[chain_id]/[address]` 

```typescript
import { NextRequest, NextResponse } from "next/server"
export const dynamic = 'force-dynamic' // defaults to force-static

export async function GET(req: NextRequest, context: {params: {chain_id: string, address: string}}) {
return NextResponse.json({
    message: "Hello, world!",
    searchParams: req.nextUrl.searchParams,
    pathname: req.nextUrl.pathname,
    chain_id: context.params.chain_id,
    address: context.params.address
  })
}
```

**Reply by IRediTOTO** [reply]
Thank you, I wish nextjs can do Types for this, without type I must GG to find it, this not cool with Next 13.
